File: frontend/earned-value-management-frontend/src/app/core/services/project.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { lastValueFrom } from 'rxjs';
import { environment } from '../../../environments/environment';
import { Project, ProjectDetail, ProjectRequest } from '../models/project.model';

@Injectable({ providedIn: 'root' })
export class ProjectService {
  private readonly http = inject(HttpClient);
  private readonly baseUrl = `${environment.apiUrl}/projects`;

  findAll(): Promise<Project[]> {
    return lastValueFrom(this.http.get<Project[]>(this.baseUrl));
  }

  findById(id: number): Promise<ProjectDetail> {
    return lastValueFrom(this.http.get<ProjectDetail>(`${this.baseUrl}/${id}`));
  }

  create(request: ProjectRequest): Promise<Project> {
    return lastValueFrom(this.http.post<Project>(this.baseUrl, request));
  }

  update(id: number, request: ProjectRequest): Promise<Project> {
    return lastValueFrom(this.http.put<Project>(`${this.baseUrl}/${id}`, request));
  }

  delete(id: number): Promise<void> {
    return lastValueFrom(this.http.delete<void>(`${this.baseUrl}/${id}`));
  }
}

File: backend/evm/src/main/java/com/earnedvaluemanagement/evm/service/ProjectService.java
package com.earnedvaluemanagement.evm.service;

import com.earnedvaluemanagement.evm.dto.request.ProjectRequest;
import com.earnedvaluemanagement.evm.dto.response.ProjectDetailResponse;
import com.earnedvaluemanagement.evm.dto.response.ProjectResponse;

import java.util.List;

public interface ProjectService {

    List<ProjectResponse> findAll();

    ProjectDetailResponse findById(Long id);

    ProjectResponse create(ProjectRequest request);

    ProjectResponse update(Long id, ProjectRequest request);

    void delete(Long id);
}
